### PR TITLE
Merges PR from jascam on Issue-1902: "Race condition in PluginCredentialProvider"

### PR DESCRIPTION
**Pull request by deepakaravindr**
Merges PR: https://github.com/NuGet/NuGet.Client/pull/196 from @jascam  on Issue-1902: "Race condition in PluginCredentialProvider".
In addition to the PR from @jascam, added logic to dispose the linkedTokenSource in a background Task 
only after all the pending tasks have completed. Note that the linkedTokenSource.Cancel() is still called
as soon as one package source returned with the package successfully. So, all the pending tasks should complete soon enough.

@yishaigalatzer @jascam @emgarten

### CODE REVIEW

File: src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs [hunk 1]
```diff
@@ -49,62 +49,78 @@ public static class PackageDownloader
...
+                    otherGroup.Add(source);
                 }
+            }
+
+            groups.Enqueue(localGroup);
+            groups.Enqueue(otherGroup);
+
+            while (groups.Count > 0)
+            {
+                token.ThrowIfCancellationRequested();
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs]: This line will prevent the dispose from ever happening if cancellation happened earlier, and still having a memory leak around

Comment (deepakaravindr) [inline src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs]: I will move it to a try-finally

File: src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs [hunk 2]
```diff
@@ -49,62 +49,78 @@ public static class PackageDownloader
...
+                            // NOTE: Create a Task out of remainingTasks which waits for all the tasks to complete
+                            // and disposes the linked token source safely. One of the tasks could try to access its
+                            // incoming CancellationToken to register a callback. If the linkedTokenSource was
+                            // disposed before being accessed, it will throw an ObjectDisposedException.
+                            // At the same time, we do not want to wait for all the tasks to complete before
+                            // disposing the linked token source. So, we start off a new Task which waits for all
+                            // the tasks to complete and then disposes the LinkedTokenSource.
+                            var remainingTasks = Task.Run(async () =>
+                            {
+                                await Task.WhenAll(tasks);
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs]: This needs a try/finally incase WhenAll throws

**Comment by emgarten** [comment]
:shipit:

**Comment by deepakaravindr** [comment]
Merged with https://github.com/NuGet/NuGet.Client/commit/07a47798efeabc7ce436d6c89f89834cabdb250d

**Pull request closed without merging**